Tools/wasm/wasm_build.py

```diff
@@ -0,0 +1,478 @@
+"""
+import argparse
+import enum
+import dataclasses
+import os
+import pathlib
+import shlex
+import shutil
+import subprocess
+import sysconfig
+from typing import Callable, List, Union  # for Python 3.8
+
+SRCDIR = pathlib.Path(__file__).parent.parent.parent.absolute()
+WASMTOOLS = SRCDIR / "Tools" / "wasm"
+BUILDDIR = SRCDIR / "builddir"
+CONFIGURE = SRCDIR / "configure"
+SETUP_LOCAL = SRCDIR / "Modules" / "Setup.local"
+
+HAS_CCACHE = shutil.which("ccache") is not None
+
+# path to WASI-SDK root
+WASI_SDK_PATH = pathlib.Path(os.environ.get("WASI_SDK_PATH", "/opt/wasi-sdk"))
+# path to Emscripten directory (upstream/emscripten subdirectory in the EMSDK)
+EMSCRIPTEN_PATH = pathlib.Path(os.environ.get("EMSCRIPTEN", "/opt/emsdk/upstream/emscripten"))
+
+# WASM_WEBSERVER = WASMTOOLS / "wasmwebserver.py"
+
+INSTALL_EMSDK = """
+wasm32-emscripten builds need Emscripten SDK. Please follow instructions at
+https://emscripten.org/docs/getting_started/downloads.html how to install
+Emscripten tp "/opt/emsdk". Alternatively you can install the SDK in a
+different location and set the environment variable
+EMSCRIPTEN=.../upstream/emscripten (the directory with emcc and
+emconfigure scripts), e.g. with ". emsdk_env.sh".
+"""
+
+INSTALL_WASI_SDK = """
+wasm32-wasi builds need WASI SDK. Please fetch the latest SDK from
+https://github.com/WebAssembly/wasi-sdk/releases and install it to
+"/opt/wasi-sdk". Alternatively you can install the SDK in a different location
+and point the environment variable WASI_SDK_PATH to the root directory
+of the SDK. The SDK is available for Linux x86_64, macOS x86_64, and MinGW.
+"""
+
+INSTALL_WASMTIME = """
+wasm32-wasi tests require wasmtime on PATH. Please follow instructions at
+https://wasmtime.dev/ to install wasmtime.
+"""
+
+
+class MissingDependency(ValueError):
+    def __init__(self, command: str, text: str):
+        self.command = command
+        self.text = text
+
+    def __str__(self):
+        return f"{type(self).__name__}: '{self.command}'\n{self.text}"
+
+
+@dataclasses.dataclass
+class Platform:
+    """Platform-specific settings
+
+    - CONFIG_SITE override
+    - configure wrapper (e.g. emconfigure)
+    - make wrapper (e.g. emmake)
+    - additional environment variables
+    - check function to verify SDK
+    """
+
+    name: str
+    pythonexe: str
+    config_site: pathlib.Path
+    configure_wrapper: pathlib.Path
+    make_wrapper: pathlib.Path
+    environ: dict
+    check: Callable[[], None]
+
+    def getenv(self) -> dict:
+        return self.environ.copy()
+
+
+NATIVE = Platform(
+    "native",
+    pythonexe=sysconfig.get_config_var("PYTHON"),  # TODO: check macOS
+    config_site=None,
+    configure_wrapper=None,
+    make_wrapper=None,
+    environ={},
+    check=lambda: None,
+)
+
+
+def _check_emscripten():
+    emconfigure = EMSCRIPTEN_PATH / "emconfigure"
+    if not emconfigure.exists():
+        raise MissingDependency(emconfigure, INSTALL_EMSDK)
```

Comment:
Might be worth doing a version check here, too (if a minimum required version is known).

Reply:
I have added a version check.